pkg/operator/ceph/csi/secrets.go

```diff
@@ -212,6 +304,123 @@ func CreateCSISecrets(context *clusterd.Context, clusterInfo *client.ClusterInfo
+// keys like `client.csi-rbd-node.1`, `client.csi-rbd-node.3`, `client.csi-rbd-node.4` so the list will have all these key name.
+// And from the list of key name we'll read the suffix index eg; for key `client.csi-rbd-node.3` we'll take `3` and compare with
+// KeyGeneration and basaed on the comparison we'll return bool to check if we should rotate the keys or not and returning list
+// of all the matching keys.
+func getCsiKeyRotationInfo(context *clusterd.Context, clusterInfo *client.ClusterInfo, clusterSpec *cephv1.ClusterSpec, keyBaseName string) (bool, []string, error) {
+	authList, err := client.AuthList(context, clusterInfo)
+	if err != nil {
+		return false, []string{}, err
+	}
+
+	KeyWithBaseName := []string{}
+	for _, entry := range authList.AuthDump {
+		if strings.Contains(entry.Entity, keyBaseName) {
+			KeyWithBaseName = append(KeyWithBaseName, entry.Entity)
+		}
+	}
+
+	if len(KeyWithBaseName) == 0 {
+		return true, KeyWithBaseName, nil
+	}
+
+	maxCount, err := getKeySuffixNumber(KeyWithBaseName[0])
+	if err != nil {
+		return false, KeyWithBaseName, errors.Wrapf(err, "failed to get currentMaxKeyCount from key entity")
+	}
+
+	currentMaxKeyCount := uint32(maxCount) //nolint:gosec // disable G115 // already checked
+	interpretedCephxStatus := cephv1.CephxStatus{
+		KeyCephVersion: "", // key ceph ver irrelevant for CSI keys
+		KeyGeneration:  uint32(currentMaxKeyCount),
+	}
+
+	// In case of CSI or overlapping key rotation, we don't need desired cephVersion as key rotation on `WithCephVersionUpdate` is not supported.
+	// We can pass any cephVersion as a place holder to the `ShouldRotateCephxKeys`.
+	shouldRotate, err := keyring.ShouldRotateCephxKeys(clusterSpec.Security.CephX.CSI.CephxConfig, clusterInfo.CephVersion, clusterInfo.CephVersion, interpretedCephxStatus)
+	if err != nil {
+		return false, KeyWithBaseName, errors.Wrap(err, "failed to determine if cephx keys should be rotated")
+	}
+
+	return shouldRotate, KeyWithBaseName, nil
+}
+
+// deleteOldKeyGen deletes the CephxKeys based on the highest KeyGenCount and keepPriorCount. We'll sort the list of strings
+// based on highest suffix in the CephXkeys. Example; `client.csi-rbd-node.3` we'll compare `3` with keepCount let's says `2`,
+// then we'll delete 1 extra CephXkeys basically deleting the last key from the sorted key name list.
+func deleteOldKeyGen(context *clusterd.Context, clusterInfo *client.ClusterInfo, allKeyWithBaseName []string, count int) error {
+	sortedClientList, err := sortCsiClientName(allKeyWithBaseName)
+	if err != nil {
+		return errors.Wrapf(err, "failed to get sorted CSI key list from %s", allKeyWithBaseName)
+	}
+
+	listToDeleteKey := []string{}
+
+	keyDeleteCount := max(0, len(sortedClientList)-count) // can't delete negative count
+	for i := range keyDeleteCount {                       // Delete the lowest-numbered keys first
+		i, err := getKeySuffixNumber(sortedClientList[i])
+		if err != nil {
+			return err
+		}
+		if i < count {
```

Comment:
Is this `i < count` check necessary?

by using `max(0, len(sortedClientList)-count)` above, we should only get a range that is exactly the keys we want to delete.

Reply:
right, changes done